articles/javascript-higher-order-functions/index.md

```diff
@@ -0,0 +1,333 @@
+
+Let's digest this information with examples.
+
+Assume:
+1. An array with domestic animals such as
+
+```js
+["cat", "dog", "cow" ,"goat", "sheep", "donkey", "pig", "horse"]
+```
+
+We want to filter all domestic animals whose names have three-letter.
+
+2. An array of adult ages.
+
+```js
+[{name: "Xavier", age: 19},
+{name: "Alice", age: 21},
+{name: "John", age: 14},
+{name: "David", age: 22},
+{name: "Amanda", age: 18},
+{name: "Amy", age: 20},
+{name: "Jesicca", age: 28},
+{name: "Alex", age: 15}]
+```
+
+We want to filter adult ages 18 and older.
+
+#### Example One: without `filter()` Method
+
+##### Domestic Animals
+
+```js
+const Animals = ["cat", "dog", "cow" ,"goat", "sheep", "donkey", "pig", "horse"];
+const dAnimals = [];
+// we could use a standard for loop
+for (let i = 0; i < Animals.length; i++) {
+  if (Animals[i].length == 3) {
+    dAnimals.push(Animals[i])
+  }
+}
+Or a for of loop
+console.log(dAnimals);
+```
+
+##### Adult Ages
+
+```js
+const adults = [
+{name: "Xavier", age: 19},
+{name: "Alice", age: 21},
+{name: "John", age: 14},
+{name: "David", age: 22},
+{name: "Amanda", age: 18},
+{name: "Amy", age: 20},
+{name: "Jesicca", age: 28},
+{name: "Alex", age: 15}
+];
+
+let anAdult = [];
+for(let i = 0; i < adults.length; i++) {
+ if(adults[i].age >= 18)  {
+    anAdult.push(adults[i]);
+  }
+}
+console.log(anAdult);
+```
+
+The examples above log and filter the test as we told it to. However, these methods can get cumbersome with a big app that has many adults to filter their ages or other adult elements. The examples are a bit lengthy and need a lot of code setup. We can compose this further using high order functions.
+
+#### Example Two: Using High Order Functions
+
+##### Domestic Animals
+
+```js
+const dAnimals = Animals.filter(function(animal) {
+  if(animal.length == 3) {
+    return true;
+  }
+});
+console.log(dAnimals);
+```
+
+##### Adult Ages
+
+```js
+const anAdult = adults.filter(function(adult) {
+  if(adult.age >= 18) {
+    return true;
+  }
+});
+console.log(anAdult);
+```
+
+We get the same results. But this time, our code is getting small and compact. No looping any more to iterate the input array to `anAdult.push(adults[i]);`
+
+####  Example Three: Arrow Function
+
+With ES6, it is unbelievable that we can compose this further using the arrow function.
+
+##### Adult Ages
+
+```js
+const anAdult = adults.filter(adults => adults.age >= 18);
+```
+
+In just one line of code and we get the same results.
+
+So how does `filter()` magically walk us to such satisfying results with just a few lines of code? `filter()` calls the passed callback once for every element being tested. `filter()` accepts the arguments of the callback function. This callback is what holds the value of the element being processed with the syntax below.
```

Comment:
Consider placing this para above. This explains the article, but the reader might not understand anything, just looking at the code.